File: brady/valid_parentheses.java
class Solution {
  
    // O(n) time, O(n) space
    public boolean isValid(String s) {

        Stack<Character> stack = new Stack<>();

        for (int i = 0; i < s.length(); ++i) {
            switch (s.charAt(i)) {
                case '(':
                case '{':
                case '[':
                    stack.push(s.charAt(i));
                    break;
                case ')':
                    if (stack.isEmpty() || stack.pop() != '(') return false;
                    break;
                case '}':
                    if (stack.isEmpty() || stack.pop() != '{') return false;
                    break;
                case ']':
                    if (stack.isEmpty() || stack.pop() != '[') return false;
            }
        }

        return stack.isEmpty();
    }
}

File: manraj/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []

        for c in s:
            if c == "(" or c == "[" or c == "{": stack.append(c)
            else:
                if not stack: return False
                pop = stack.pop()
                if c == ")" and pop != "(": return False
                if c == "]" and pop != "[": return False
                if c == "}" and pop != "{": return False
        
        return not stack
            

